### PR TITLE
Support for quota_project in credential.

### PR DESCRIPTION
**Pull request by amanda-tarafa**

### CODE REVIEW

**Review by jskeet — COMMENTED** [review]
This one definitely feels like we need to consult Jan pretty soon. I've a feeling we might want to create multiple CallCredentials and combine them, but I do see that it's going to get messy if we're *given* call credentials and we don't know what's specifying a project and what's not.

(I think we also need clarification from Yoshi about the design. This feels like non-auth info that comes across *with* credentials, but it isn't really *part* of a credential.)

**Comment by amanda-tarafa** [comment]
I'm still torn about it feeling like non-auth info, that was definetely my first inclination, but thinking more about it, I'm not so sure anymore. For resource based APIs, the project owning the resource involved in the API call takes precedence over anything we send for quotas and billing. But for non resource based APIs, this can be seen effectively as an impersonation mechanism, which is definetely auth. In which case I even think that both values should be set on the same interceptor (and just one credential) since the credential should effectively be the result of the token + project.
Which gets me now to backtrack about having a general interface ITokenAccessWithExtraInfo (because that info can be more random in nature) in favour of IImpersonatorTokenAccess (for lack of a better name) with the one property for the project.

File: src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs [hunk 1]
```diff
@@ -43,29 +45,22 @@ public static class GoogleAuthInterceptors
...
-        /// </summary>
-        /// <param name="credential">The credential to use to obtain access tokens.</param>
-        /// <returns>The interceptor.</returns>
-        public static AsyncAuthInterceptor FromCredential(ITokenAccessWithExtraInformation credential)
-        {
-            return new AsyncAuthInterceptor(async (context, metadata) => {
-                await AddBearerTokenHeader(credential, context, metadata).ConfigureAwait(false);
-                AddExtraInforHeaders(credential, context, metadata);
-            });
+            if (credential is IProjectImpersonatorTokenAccess impersonator)
```
Comment (amanda-tarafa) [inline src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs:54]: Somewhat dirty implementation that keeps the surface cleaner. If we need to add more similar info in the future, we can add more interfaces in Google.Apis.Auth and the corresponding ifs here.

**Review by amanda-tarafa — COMMENTED** [review]
@jskeet I've added one complementing commit in each repo with a different implementation proposal based on some of our comments.
Please ignore the extra commits in this repo, I rebased out of habit.

There are now two commits here that are relevant:
 - [grpc/ITokenAccessWithExtraInfo](https://github.com/amanda-tarafa/grpc/pull/1/commits/e8b0c57253db529813ce0d33a040a49883e0d35a) which complements [google-api-dotnet-client/ITokenAccessWithExtraInfo](https://github.com/amanda-tarafa/google-api-dotnet-client/pull/1/commits/ef2fe5a184646e40cbd80f2befc90dcd33acbbab). These you already saw.
- [grpc/IProjectImpersonatorTokenAccess](https://github.com/amanda-tarafa/grpc/pull/1/commits/464a4003beffbcd512443576b1f0cd9af9b45502) which compliments [google-api-dotnet-client/IProjectImpersonatorTokenAccess](https://github.com/amanda-tarafa/google-api-dotnet-client/pull/1/commits/a8201403a093fc9c9c623e7d7c1de32cc8a55bdd). These are the new ones and I prefer them now.

Let me know.

**Comment by jskeet** [comment]
Right - will try to have a look later this afternoon.

**Comment by jtattermusch** [comment]
Nit: something's wrong with the commit history of this PR (there's a lot of unrelated junk. Perhaps you should update master in your fork to get a cleaner history?

File: src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs [hunk 2]
```diff
@@ -43,29 +45,22 @@ public static class GoogleAuthInterceptors
...
+            {
+                interceptors += new AsyncAuthInterceptor((context, metadata) => 
+                {
+                    metadata.Add(CreateHeader(QuotaProjectIdHeader, impersonator.QuotaProjectId));
```
Comment (jtattermusch) [inline src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs:58]: My general concern about `IProjectImpersonatorTokenAccess` is that it's limited to a single functionality and header name. Problems with that:
- if more headers like this are added (could always happen), we will need to make the same hack again
- like this we need to duplicate the Google auth logic in Grpc.Auth (remember the name of the header, a specific interface etc.)

Comment (amanda-tarafa) [reply src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs:58]: Yep, this I don't like either.
On the other hand with the ITokenAccessWithExtraInformation implementation we are risking information not related to auth piggy backing on the token.
I don't know if you saw [the ITokenAccessWithExtraInformation](https://github.com/amanda-tarafa/grpc/pull/1/commits/fb5a2992740703beeea972369c26f181f8b906ff) commit here, given that I had the dirty commits on the PR (I already fixed that). If you haven't, take a look at that proposal and let me know if you prefer it.

**Review by amanda-tarafa — COMMENTED** [review]
Fixed the PR history and replied. Thanks!

File: src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs (hunk 2 above)
Comment (jtattermusch) [reply src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs:58]: Yes, I saw the other proposal (ITokenAccessWithExtraInformation) and I like it better (as described in the other PR).

File: src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs [hunk 3]
```diff
@@ -43,10 +43,28 @@ public static class GoogleAuthInterceptors
         /// <returns>The interceptor.</returns>
         public static AsyncAuthInterceptor FromCredential(ITokenAccess credential)
         {
+            if(credential is ITokenAccessWithExtraInformation credentialWithExtra)
```
Comment (jtattermusch) [inline src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs]: I think the overall design for ITokenAccessWithExtraInformation makes sense (not considering the details of the implementation).
The `credential is ITokenAccessWithExtraInformation credentialWithExtra` approach makes sense but the danger is that if one accidentally has an old version of Grpc.Auth (without these changes), the ITokenAccessWithExtraInformation object will be viewed as just a regular ITokenAccess and the extra headers will simply be ignored silently (and I can imagine users not knowing what's going on and filing issues that the new funtionality doesn't work). Have we considered that risk?

Comment (amanda-tarafa) [reply src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs]: I think this is the same as if one accidentally has an older version of Google.Apis.Auth, in which case the quota_project field in Json will be ignored when creating the token.
Either way, as you say, we will be silently ignoring the header. Server might or might not fail whether it was expecting the header for the credential or not (tbh I'm not sure about that). At worst, the operation will be billed on a different project than intended, but nonetheless a project that the token was authorized to bill.
I think this is something we can't control on the client library side. I will make sure to raise this internally.

**Comment by amanda-tarafa** [comment]
Closing as I will start the real PR now.

**Pull request closed without merging**